### PR TITLE
Adding mikkt tangent support to CSG objects

### PR DESCRIPTION
**Pull request by BastiaanOlij**
CSG objects now calculate tangents so normal mapping will work. Most CSG objects don't calculate very nice UVs so not sure how useful this is but it does work very well with CSGMesh and CSGPolygon.

Here is an example of using two CSGMesh objects in combination with a cube and sphere mesh primitive:
image

### CODE REVIEW

File: modules/csg/csg_shape.cpp [hunk 1]
```diff
@@ -159,6 +159,74 @@ CSGBrush *CSGShape::_get_brush() {
 	return brush;
 }
 
+int CSGShape::mikktGetNumFaces(const SMikkTSpaceContext *pContext) {
```
Comment (akien-mga) [inline modules/csg/csg_shape.cpp]: Reminder that we use `snake_case` for identifiers in Godot's codebase :)
So it should be `mikkt_get_num_faces` and `p_context`.

Comment (BastiaanOlij) [reply modules/csg/csg_shape.cpp]: @akien-mga those are the names already used in surface tool but I'm happy to change them

**Comment by BastiaanOlij** [comment]
Request from @reduz was to make it possible to turn off the tangent calculation if its unwanted as it can eat up performance. It will be enabled by default and only applied on the final mesh. Commit to follow

File: modules/csg/csg_shape.cpp (hunk 1 above)
Comment (akien-mga) [reply modules/csg/csg_shape.cpp]: I see - it's fine to keep them as is for consistency with `SurfaceTool` then. I guess those were used in ST for easy reference to the thirdparty mikktspace API.

**Comment by Toshiwoz** [comment]
imagen
Hey, I'm not sure if it has to do with this but, if I use an ArrayMesh I generate from code, i get the normals as in the picture, quite strange as in a regular Mesh does not do that.
imagen
[terrain.zip](https://github.com/godotengine/godot/files/2609141/terrain.zip)
It can be because the mesh is composed of 16 surfaces? Or the way I built it, not sure.

Should I open a separate issue?

**Comment by BastiaanOlij** [comment]
@Toshiwoz , this looks like a difference between smooth and non-smooth shading.

**Comment by BastiaanOlij** [comment]
Owh and yes, raise a new issue I'd say or it'll get lost ;)

**Pull request merged**